### PR TITLE
First step towards cluster-dump command

### PR DESCRIPTION
**Pull request by justinsb**
Can dump a VPC

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1274)

### CODE REVIEW

File: cmd/kops/toolbox.go [hunk 1]
```diff
@@ -18,14 +18,18 @@ package main
```
Comment (chrislovecnm) [inline cmd/kops/toolbox.go]: Is toolbox the right place?  I see this option getting kinda big, and may even be a container running on k8s to push logs to s3.

File: cmd/kops/toolbox_dump.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+	resources, err := d.ListResources()
+	if err != nil {
+		return err
+	}
+
+	data := make(map[string]interface{})
+
+	switch options.Output {
+	case OutputYaml:
```
Comment (chrislovecnm) [inline cmd/kops/toolbox_dump.go]: Should we look at json as well?

Comment (chrislovecnm) [inline cmd/kops/toolbox_dump.go]: maybe a issue item rather than putting it here.

File: upup/pkg/kutil/delete_cluster.go [hunk 3]
```diff
@@ -1213,6 +1215,14 @@ func DeleteVPC(cloud fi.Cloud, r *ResourceTracker) error {
 	return nil
 }
 
+func DumpVPC(r *ResourceTracker) (interface{}, error) {
```
Comment (chrislovecnm) [inline upup/pkg/kutil/delete_cluster.go]: Dumb question.  What is this in delete_cluster?

File: cmd/kops/toolbox.go [hunk 4]
```diff
@@ -18,14 +18,18 @@ package main
...
+	"io"
+	"k8s.io/kops/cmd/kops/util"
 )
 
-// toolboxCmd represents the toolbox command
-var toolboxCmd = &cobra.Command{
-	Use:   "toolbox",
-	Short: "Misc infrequently used commands",
-}
+func NewCmdToolbox(f *util.Factory, out io.Writer) *cobra.Command {
```
Comment (kris-nova) [inline cmd/kops/toolbox.go]: Don't love the name - but you guys know I am a fan of concrete verbose names over misc naming.. 

Either way, I think we should have a decent chunk of documentation around this. Especially if we are dropping the cluster dump into a `toolbox` command

File: cmd/kops/toolbox_dump.go (hunk 2 above)
Comment (kris-nova) [inline cmd/kops/toolbox_dump.go]: Yeah let's open an issue. Looks like we are already switching on output type which is good, just need to flip the encoding at the end of the logic.

File: upup/pkg/kutil/delete_cluster.go (hunk 3 above)
Comment (kris-nova) [inline upup/pkg/kutil/delete_cluster.go]: Looks like we are using the `delete` cluster for something here.. which is awesome..

Right now kops is able to find cluster resources to delete by going through and querying AWS for tags, and resource names.. So it seems like a lot of the cluster dump overhead might already be here.. my only thought on the matter is that the cluster definition (dump) should come **before** the delete.

In other words..

We have the cluster dump (a way to poll AWS for resources) that can be used in many ways, and delete is simply a wrapper that uses the dump, as well as outputting the dump to json, yaml, etc..

Just a thought 😄

File: cmd/kops/toolbox.go (hunk 1 above)
Comment (justinsb) [inline cmd/kops/toolbox.go]: It's a good place to start I think; I think it's reasonable to move it out of the toolbox later (I feel that is more acceptable in the toolbox than elsewhere)

File: cmd/kops/toolbox_dump.go (hunk 2 above)
Comment (justinsb) [inline cmd/kops/toolbox_dump.go]: Right, just me starting with MVP

File: upup/pkg/kutil/delete_cluster.go (hunk 3 above)
Comment (justinsb) [inline upup/pkg/kutil/delete_cluster.go]: That's right.  We should refactor discovery from deletion.

There's also another use case - right now if we switch e.g. from private -> public, we're likely to orphan a bunch of resources.  We have no way to delete them right now.  We should have some form of garbage collection.  Although I guess that is more similar to delete :-)

File: cmd/kops/toolbox.go (hunk 4 above)
Comment (justinsb) [inline cmd/kops/toolbox.go]: I'm hiding it in the toolbox until we figure out where to put it :-)

File: cmd/kops/toolbox_dump.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+			o, err := r.Dumper(r)
+			if err != nil {
+				return fmt.Errorf("error dumping %q: %v", k, err)
+			}
+			if o != nil {
+				dumpedResources = append(dumpedResources, o)
+			}
+		}
+		data["resources"] = dumpedResources
+		b, err := kops.ToRawYaml(data)
```
Comment (justinsb) [inline cmd/kops/toolbox_dump.go]: We should.  I think we can actually, just starting small!

**Comment by chrislovecnm** [comment]
@kris-nova thoughts?

**Comment by justinsb** [comment]
Assigning to you @kris-nova as I don't think @chrislovecnm has remaining objections (?)

**Pull request merged**